bug: .tool-versions in home is not applied as global

### Describe the Bug

I had to apply .tool-versions to / and not to ~/ - "/home/user"

To fix this I copied .tool-versions from ~/ to /, making it the same as the global version

### Steps to Reproduce

rodar algum comando em um nivel fora da home 

cd ..




### Expected Behaviour

Recognize ~/.tool-versions as the system-wide default

### Actual Behaviour

is not logging in lower level folder outside ~/ so accepting commands inside /home/user

### Environment

```shell
alvaroico in ~
❯ asdf info
OS:
Linux Bilu 5.15.167.4-microsoft-standard-WSL2 #1 SMP Tue Nov 5 00:21:55 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

SHELL:
zsh 5.9 (x86_64-ubuntu-linux-gnu)

BASH VERSION:
5.2.21(1)-release

ASDF VERSION:
v0.16.0

ASDF INTERNAL VARIABLES:
ASDF_DEFAULT_TOOL_VERSIONS_FILENAME=.tool-versions
ASDF_DATA_DIR=/home/alvaroico/.asdf
ASDF_CONFIG_FILE=/home/alvaroico/.asdfrc

ASDF INSTALLED PLUGINS:
bun         https://github.com/cometkim/asdf-bun.git              019058c3a9f408a06bab63a4a3782c555a8dde3c
deno        https://github.com/asdf-community/asdf-deno.git       13f7f08c93989fdc0b372e78df226e511fc13fa9
dotnet-core https://github.com/emersonsoares/asdf-dotnet-core.git 4583e4b87b48d39e37ed0acbf54481297d6b4992
elixir      https://github.com/asdf-vm/asdf-elixir.git            ea842831ae39417f2dfe7b96fb0c3cc573314aa6
erlang      https://github.com/asdf-vm/asdf-erlang.git            0e617b61ece7cff73331aacc1cd9986e35618c41
flutter     https://github.com/oae/asdf-flutter.git               d73606641938341b8b17bd3e2859a5267bb89bd0
golang      https://github.com/asdf-community/asdf-golang.git     e2527a31714da7bc671a684308579f4ef8863281
java        https://github.com/halcyon/asdf-java.git              7c7ce22c2dc1ce084aa23d6aec60698e73dc9046
nodejs      https://github.com/asdf-vm/asdf-nodejs.git            93bd217ba74f05080eebb7d8e1c146a385b01d7f
php         https://github.com/asdf-community/asdf-php.git        52c669a77f08ec81caff47d560476d2ba1b176b9
python      https://github.com/danhper/asdf-python.git            a3a01856098d6d2b9642e382f5b38e70275726d1
ruby        https://github.com/asdf-vm/asdf-ruby.git              059cdee5b02c5a5569a5858e2a379f84dbdb95d7
rust        https://github.com/code-lever/asdf-rust.git           95acf4fe65df1de74fca502482b8f3ac5af73c05
alvaroico in ~
❯ cd ..
alvaroico in /home 
❯ asdf info
OS:
Linux Bilu 5.15.167.4-microsoft-standard-WSL2 #1 SMP Tue Nov 5 00:21:55 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

SHELL:
zsh 5.9 (x86_64-ubuntu-linux-gnu)

BASH VERSION:
5.2.21(1)-release

ASDF VERSION:
v0.16.0

ASDF INTERNAL VARIABLES:
ASDF_DEFAULT_TOOL_VERSIONS_FILENAME=.tool-versions
ASDF_DATA_DIR=/home/alvaroico/.asdf
ASDF_CONFIG_FILE=/home/alvaroico/.asdfrc

ASDF INSTALLED PLUGINS:
bun         https://github.com/cometkim/asdf-bun.git              019058c3a9f408a06bab63a4a3782c555a8dde3c
deno        https://github.com/asdf-community/asdf-deno.git       13f7f08c93989fdc0b372e78df226e511fc13fa9
dotnet-core https://github.com/emersonsoares/asdf-dotnet-core.git 4583e4b87b48d39e37ed0acbf54481297d6b4992
elixir      https://github.com/asdf-vm/asdf-elixir.git            ea842831ae39417f2dfe7b96fb0c3cc573314aa6
erlang      https://github.com/asdf-vm/asdf-erlang.git            0e617b61ece7cff73331aacc1cd9986e35618c41
flutter     https://github.com/oae/asdf-flutter.git               d73606641938341b8b17bd3e2859a5267bb89bd0
golang      https://github.com/asdf-community/asdf-golang.git     e2527a31714da7bc671a684308579f4ef8863281
java        https://github.com/halcyon/asdf-java.git              7c7ce22c2dc1ce084aa23d6aec60698e73dc9046
nodejs      https://github.com/asdf-vm/asdf-nodejs.git            93bd217ba74f05080eebb7d8e1c146a385b01d7f
php         https://github.com/asdf-community/asdf-php.git        52c669a77f08ec81caff47d560476d2ba1b176b9
python      https://github.com/danhper/asdf-python.git            a3a01856098d6d2b9642e382f5b38e70275726d1
ruby        https://github.com/asdf-vm/asdf-ruby.git              059cdee5b02c5a5569a5858e2a379f84dbdb95d7
rust        https://github.com/code-lever/asdf-rust.git           95acf4fe65df1de74fca502482b8f3ac5af73c05
alvaroico in /home 
```

### asdf plugins affected (if relevant)

![Image](https://github.com/user-attachments/assets/8a95ce69-ef47-45f7-9a4e-5b9b903e9547)

Hi there! I have the same kind of error when installing Ruby (which uses Rust to compile it's internal JIT) : 

```sh
building Rust YJIT (release mode)
No version is set for command rustc
Consider adding one of the following versions in your config file at /tmp/ruby-build.20250130213132.2471.si17Lg/ruby-3.4.1/.tool-versions
rust 1.83
make[1]: *** [yjit/yjit.mk:26: yjit/target/release/libyjit.a] Error 126
```

```sh
OS:
Linux cloud 6.1.0-30-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.124-1 (2025-01-12) x86_64 GNU/Linux

SHELL:
GNU bash, version 5.2.15(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

BASH VERSION:
5.2.15(1)-release

ASDF VERSION:
v0.16.0

ASDF INTERNAL VARIABLES:
ASDF_DEFAULT_TOOL_VERSIONS_FILENAME=.tool-versions
ASDF_DATA_DIR=/data/nicolas/.asdf
ASDF_CONFIG_FILE=/data/nicolas/.asdfrc

ASDF INSTALLED PLUGINS:
bombardier https://github.com/NeoHsu/asdf-bombardier.git          46bea4d29d44a1578f7114bef0c86bfe9bbc3ee9
crystal    https://github.com/asdf-community/asdf-crystal.git     7af37cf7b63ad0c3c176a1e740b2d0919f41f054
earthly    https://github.com/YR-ZR0/asdf-earthly.git             58c78cb8b7b61e5b68f00714d0c8739f93cb5590
golang     https://github.com/asdf-community/asdf-golang.git      e2527a31714da7bc671a684308579f4ef8863281
k3d        https://github.com/spencergilbert/asdf-k3d.git         5b0e5b9209bf0e1e8f83cbd912e8cccfe31e7fe2
kompose    https://github.com/technikhil314/asdf-kompose.git      2706fa5579473c50aeb79c6098c7797ea96eb64c
kubectl    https://github.com/asdf-community/asdf-kubectl.git     2fb3b57090e3efe55953a751a134c271ba83b119
mkcert     https://github.com/salasrod/asdf-mkcert.git            8287d3c115b6c681592f44b2060973af6ae24be3
nodejs     https://github.com/asdf-vm/asdf-nodejs.git             93bd217ba74f05080eebb7d8e1c146a385b01d7f
overmind   https://github.com/virtualstaticvoid/asdf-overmind.git abaee0622091883c16f25cf860f19b0f3a2b639c
python     https://github.com/danhper/asdf-python.git             a3a01856098d6d2b9642e382f5b38e70275726d1
ruby       https://github.com/asdf-vm/asdf-ruby.git               059cdee5b02c5a5569a5858e2a379f84dbdb95d7
rust       https://github.com/code-lever/asdf-rust.git            95acf4fe65df1de74fca502482b8f3ac5af73c05
shellcheck https://github.com/luizm/asdf-shellcheck.git           66200ffa221fc75d7b28103f53326db6e721d08f
skaffold   https://github.com/nklmilojevic/asdf-skaffold.git      bf2f9922ddfc4d8613a2d92f3fedc458464e36e6
squarectl  https://github.com/jbox-web/asdf-squarectl.git         5c0621cf4d1402c06cff70734c22a8c0f227198c
tokei      https://github.com/gasuketsu/asdf-tokei.git            d127410f0e4acf56b2e8070108df8a3847047fc6
yarn       https://github.com/twuni/asdf-yarn.git                 376c5407ef22c82fa56e4bd9f6a9108354063acb
yq         https://github.com/sudermanjr/asdf-yq.git              772992f66c49fe681b1cd02d14e8166930472bf1
```

Rust 1.83 is declared globally :

```sh
asdf list rust
 *1.83
```

```sh
rustc --version
rustc 1.83.0 (90b35a623 2024-11-26)
```

So, i just tried to set add the shims folder to my path and that fixed it.

```
~
> export PATH="$HOME/.asdf/shims:${PATH}"

~
> mvn
[INFO] Scanning for projects...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  0.077 s
[INFO] Finished at: 2025-01-31T16:23:39+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] No goals have been specified for this build. You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: pre-clean, clean, post-clean, validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-site, site, post-site, site-deploy. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/NoGoalSpecifiedException

~
> java --version
java 21.0.2 2024-01-16 LTS
Java(TM) SE Runtime Environment (build 21.0.2+13-LTS-58)
Java HotSpot(TM) 64-Bit Server VM (build 21.0.2+13-LTS-58, mixed mode, sharing)

~
> mvn --version
Apache Maven 3.9.8 (36645f6c9b5079805ea5009217e36f2cffd34256)
Maven home: /home/eriase/.asdf/installs/maven/3.9.8
Java version: 21.0.2, vendor: Oracle Corporation, runtime: /home/eriase/.asdf/installs/java/oracle-21.0.2
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.15.167.4-microsoft-standard-wsl2", arch: "amd64", family: "unix"
```

--- a/internal/resolve/resolve.go
+++ b/internal/resolve/resolve.go
@@ -36,7 +36,17 @@ func Version(conf config.Config, plugin plugins.Plugin, directory string) (versi
 		}
 
 		nextDir := path.Dir(directory)
+		// If current dir and next dir are the same it means we've reached `/` and
+		// have no more parent directories to search.
 		if nextDir == directory {
+			// If no version found, try current users home directory. I'd like to
+			// eventually remove this feature.
+			homeDir, osErr := os.UserHomeDir()
+			if osErr != nil {
+				break
+			}
+
+			versions, found, err = findVersionsInDir(conf, plugin, homeDir)
 			break
 		}
 		directory = nextDir

--- a/internal/resolve/resolve_test.go
+++ b/internal/resolve/resolve_test.go
@@ -4,6 +4,7 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"strings"
 	"testing"
 
 	"github.com/asdf-vm/asdf/internal/config"
@@ -12,13 +13,15 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+const testPluginName = "test-plugin"
+
 func TestVersion(t *testing.T) {
 	testDataDir := t.TempDir()
 	currentDir := t.TempDir()
 	conf := config.Config{DataDir: testDataDir, DefaultToolVersionsFilename: ".tool-versions", ConfigFile: "testdata/asdfrc"}
-	_, err := repotest.InstallPlugin("dummy_plugin", conf.DataDir, "lua")
+	_, err := repotest.InstallPlugin("dummy_plugin", conf.DataDir, testPluginName)
 	assert.Nil(t, err)
-	plugin := plugins.New(conf, "lua")
+	plugin := plugins.New(conf, testPluginName)
 
 	t.Run("returns empty slice when non-existent version passed", func(t *testing.T) {
 		toolVersion, found, err := Version(conf, plugin, t.TempDir())
@@ -29,7 +32,7 @@ func TestVersion(t *testing.T) {
 
 	t.Run("returns single version from .tool-versions file", func(t *testing.T) {
 		// write a version file
-		data := []byte("lua 1.2.3")
+		data := []byte(fmt.Sprintf("%s 1.2.3", testPluginName))
 		err = os.WriteFile(filepath.Join(currentDir, ".tool-versions"), data, 0o666)
 
 		toolVersion, found, err := Version(conf, plugin, currentDir)
@@ -40,10 +43,10 @@ func TestVersion(t *testing.T) {
 
 	t.Run("returns version from env when env variable set", func(t *testing.T) {
 		// Set env
-		t.Setenv("ASDF_LUA_VERSION", "2.3.4")
+		t.Setenv(fmt.Sprintf("ASDF_%s_VERSION", strings.ToUpper(testPluginName)), "2.3.4")
 
 		// write a version file
-		data := []byte("lua 1.2.3")
+		data := []byte(fmt.Sprintf("%s 1.2.3", testPluginName))
 		err = os.WriteFile(filepath.Join(currentDir, ".tool-versions"), data, 0o666)
 
 		// assert env variable takes precedence
@@ -55,7 +58,7 @@ func TestVersion(t *testing.T) {
 
 	t.Run("returns single version from .tool-versions file in parent directory", func(t *testing.T) {
 		// write a version file
-		data := []byte("lua 1.2.3")
+		data := []byte(fmt.Sprintf("%s 1.2.3", testPluginName))
 		err = os.WriteFile(filepath.Join(currentDir, ".tool-versions"), data, 0o666)
 
 		subDir := filepath.Join(currentDir, "subdir")
@@ -72,9 +75,9 @@ func TestVersion(t *testing.T) {
 func TestFindVersionsInDir(t *testing.T) {
 	testDataDir := t.TempDir()
 	conf := config.Config{DataDir: testDataDir, DefaultToolVersionsFilename: ".tool-versions", ConfigFile: "testdata/asdfrc"}
-	_, err := repotest.InstallPlugin("dummy_plugin", conf.DataDir, "lua")
+	_, err := repotest.InstallPlugin("dummy_plugin", conf.DataDir, testPluginName)
 	assert.Nil(t, err)
-	plugin := plugins.New(conf, "lua")
+	plugin := plugins.New(conf, testPluginName)
 
 	t.Run("when no versions set returns found false", func(t *testing.T) {
 		currentDir := t.TempDir()
@@ -89,7 +92,7 @@ func TestFindVersionsInDir(t *testing.T) {
 	t.Run("when version is set returns found true and version", func(t *testing.T) {
 		currentDir := t.TempDir()
 
-		data := []byte("lua 1.2.3")
+		data := []byte(fmt.Sprintf("%s 1.2.3", testPluginName))
 		err = os.WriteFile(filepath.Join(currentDir, ".tool-versions"), data, 0o666)
 
 		toolVersion, found, err := findVersionsInDir(conf, plugin, currentDir)
@@ -102,7 +105,7 @@ func TestFindVersionsInDir(t *testing.T) {
 	t.Run("when multiple versions present in .tool-versions returns found true and versions", func(t *testing.T) {
 		currentDir := t.TempDir()
 
-		data := []byte("lua 1.2.3 2.3.4")
+		data := []byte(fmt.Sprintf("%s 1.2.3 2.3.4", testPluginName))
 		err = os.WriteFile(filepath.Join(currentDir, ".tool-versions"), data, 0o666)
 
 		toolVersion, found, err := findVersionsInDir(conf, plugin, currentDir)
@@ -116,7 +119,7 @@ func TestFindVersionsInDir(t *testing.T) {
 		conf := config.Config{DataDir: testDataDir, DefaultToolVersionsFilename: "custom-file"}
 		currentDir := t.TempDir()
 
-		data := []byte("lua 1.2.3 2.3.4")
+		data := []byte(fmt.Sprintf("%s 1.2.3 2.3.4", testPluginName))
 		err = os.WriteFile(filepath.Join(currentDir, "custom-file"), data, 0o666)
 
 		toolVersion, found, err := findVersionsInDir(conf, plugin, currentDir)
@@ -143,9 +146,9 @@ func TestFindVersionsInDir(t *testing.T) {
 func TestFindVersionsLegacyFiles(t *testing.T) {
 	testDataDir := t.TempDir()
 	conf := config.Config{DataDir: testDataDir}
-	_, err := repotest.InstallPlugin("dummy_plugin", conf.DataDir, "lua")
+	_, err := repotest.InstallPlugin("dummy_plugin", conf.DataDir, testPluginName)
 	assert.Nil(t, err)
-	plugin := plugins.New(conf, "lua")
+	plugin := plugins.New(conf, testPluginName)
 
 	t.Run("when given tool that lacks list-legacy-filenames callback returns empty versions list", func(t *testing.T) {
 		pluginName := "foobar"

--- a/internal/shims/shims.go
+++ b/internal/shims/shims.go
@@ -93,7 +93,9 @@ func FindExecutable(conf config.Config, shimName, currentDirectory string) (stri
 				}
 
 				versions.Versions = tempVersions
-				existingPluginToolVersions[plugin] = versions
+				if len(versions.Versions) > 0 {
+					existingPluginToolVersions[plugin] = versions
+				}
 			}
 		}
 	}
@@ -131,9 +133,9 @@ func FindExecutable(conf config.Config, shimName, currentDirectory string) (stri
 
 	tools := []string{}
 	versions := []string{}
-	for plugin := range existingPluginToolVersions {
+	for plugin, toolVersions := range existingPluginToolVersions {
 		tools = append(tools, plugin.Name)
-		versions = append(versions, existingPluginToolVersions[plugin].Versions...)
+		versions = append(versions, toolVersions.Versions...)
 	}
 
 	return "", plugins.Plugin{}, "", false, NoExecutableForPluginError{shim: shimName, tools: tools, versions: versions}

--- a/internal/versions/testdata/asdfrc
+++ b/internal/versions/testdata/asdfrc
@@ -1,4 +1,4 @@
-pre_asdf_download_lua = echo pre_asdf_download_lua $@
-pre_asdf_install_lua = echo pre_asdf_install_lua $@
-post_asdf_install_lua = echo post_asdf_install_lua $@
+pre_asdf_download_testlua = echo pre_asdf_download_lua $@
+pre_asdf_install_testlua = echo pre_asdf_install_lua $@
+post_asdf_install_testlua = echo post_asdf_install_lua $@
 always_keep_download = yes

--- a/internal/versions/versions_test.go
+++ b/internal/versions/versions_test.go
@@ -14,7 +14,7 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-const testPluginName = "lua"
+const testPluginName = "testlua"
 
 func TestInstallAll(t *testing.T) {
 	t.Run("installs multiple tools when multiple tool versions are specified", func(t *testing.T) {

